File: system-test/dotnet/SystemTests/Legacy/Mod03/E2eTests/PlaceOrderPositiveApiTest.cs
using System.Net;
using System.Text;
using System.Text.Json;
using SystemTests.Commons.Constants;
using SystemTests.Legacy.Mod03.E2eTests.Base;
using Shouldly;
using Xunit;

namespace SystemTests.Legacy.Mod03.E2eTests;

public class PlaceOrderPositiveApiTest : BaseE2eTest
{
    protected override Task SetShopRawAsync()
    {
        SetUpShopHttpClient();
        return Task.CompletedTask;
    }

    [Fact]
    public async Task ShouldPlaceOrderForValidInput()
    {
        var sku = CreateUniqueSku(Defaults.SKU);
        var createProductJson = $$"""{"id":"{{sku}}","title":"Test Product","description":"Test Description","category":"Test Category","brand":"Test Brand","price":"20.00"}""";

        var createProductUri = new Uri(_configuration.ErpBaseUrl + "/api/products");
        var createProductContent = new StringContent(createProductJson, Encoding.UTF8, "application/json");
        var createProductResponse = await _erpHttpClient!.PostAsync(createProductUri, createProductContent);
        ((int)createProductResponse.StatusCode).ShouldBe(201);

        var placeOrderJson = $$"""{"sku":"{{sku}}","quantity":"5","country":"US"}""";

        var placeOrderUri = new Uri(_configuration.ShopApiBaseUrl + "/api/orders");
        var placeOrderContent = new StringContent(placeOrderJson, Encoding.UTF8, "application/json");
        var placeOrderResponse = await _shopApiHttpClient!.PostAsync(placeOrderUri, placeOrderContent);
        ((int)placeOrderResponse.StatusCode).ShouldBe(201);

        var placeOrderBody = await placeOrderResponse.Content.ReadAsStringAsync();
        using var placeOrderDoc = JsonDocument.Parse(placeOrderBody);
        var orderNumber = placeOrderDoc.RootElement.GetProperty("orderNumber").GetString()!;
        orderNumber.ShouldStartWith("ORD-");

        var viewOrderUri = new Uri(_configuration.ShopApiBaseUrl + $"/api/orders/{orderNumber}");
        var viewOrderResponse = await _shopApiHttpClient!.GetAsync(viewOrderUri);
        ((int)viewOrderResponse.StatusCode).ShouldBe((int)HttpStatusCode.OK);

        var viewOrderBody = await viewOrderResponse.Content.ReadAsStringAsync();
        using var viewOrderDoc = JsonDocument.Parse(viewOrderBody);
        var order = viewOrderDoc.RootElement;
        order.GetProperty("orderNumber").GetString().ShouldBe(orderNumber);
        order.GetProperty("sku").GetString().ShouldBe(sku);
        order.GetProperty("quantity").GetInt32().ShouldBe(5);
        order.GetProperty("unitPrice").GetDecimal().ShouldBe(20.00m);
        order.GetProperty("basePrice").GetDecimal().ShouldBe(100.00m);
        order.GetProperty("totalPrice").GetDecimal().ShouldBeGreaterThan(0);
        order.GetProperty("status").GetString().ShouldBe("PLACED");
    }
}

File: system-test/java/src/test/java/com/optivem/eshop/systemtest/legacy/mod03/e2e/PlaceOrderPositiveApiTest.java
package com.optivem.eshop.systemtest.legacy.mod03.e2e;

import com.optivem.eshop.systemtest.legacy.mod03.e2e.base.BaseE2eTest;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

import java.net.URI;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;

import static com.optivem.eshop.systemtest.commons.constants.Defaults.*;
import static org.assertj.core.api.Assertions.assertThat;

class PlaceOrderPositiveApiTest extends BaseE2eTest {
    @Override
    protected void setShopClient() {
        setUpShopHttpClient();
    }

    @Test
    void shouldPlaceOrderWithCorrectSubtotalPrice() throws Exception {
        var sku = createUniqueSku(SKU);
        var createProductJson = """
                {
                    "id": "%s",
                    "title": "Test Product",
                    "description": "Test Description",
                    "category": "Test Category",
                    "brand": "Test Brand",
                    "price": "20.00"
                }
                """.formatted(sku);

        var createProductUri = URI.create(getErpBaseUrl() + "/api/products");
        var createProductRequest = HttpRequest.newBuilder()
                .uri(createProductUri)
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(createProductJson))
                .build();

        var createProductResponse = erpHttpClient.send(createProductRequest, HttpResponse.BodyHandlers.ofString());
        assertThat(createProductResponse.statusCode()).isEqualTo(201);

        var placeOrderJson = """
                {
                    "sku": "%s",
                    "quantity": "5"
                }
                """.formatted(sku);

        var placeOrderUri = URI.create(getShopApiBaseUrl() + "/api/orders");
        var placeOrderRequest = HttpRequest.newBuilder()
                .uri(placeOrderUri)
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(placeOrderJson))
                .build();

        var placeOrderResponse = shopApiHttpClient.send(placeOrderRequest, HttpResponse.BodyHandlers.ofString());
        assertThat(placeOrderResponse.statusCode()).isEqualTo(201);

        var placeOrderBody = httpObjectMapper.readTree(placeOrderResponse.body());
        var orderNumber = placeOrderBody.get("orderNumber").asText();

        var viewOrderUri = URI.create(getShopApiBaseUrl() + "/api/orders/" + orderNumber);
        var viewOrderRequest = HttpRequest.newBuilder()
                .uri(viewOrderUri)
                .header("Content-Type", "application/json")
                .GET()
                .build();

        var viewOrderResponse = shopApiHttpClient.send(viewOrderRequest, HttpResponse.BodyHandlers.ofString());
        assertThat(viewOrderResponse.statusCode()).isEqualTo(200);

        var viewOrderBody = httpObjectMapper.readTree(viewOrderResponse.body());
        assertThat(viewOrderBody.get("totalPrice").asDouble()).isEqualTo(100.00);
    }

    @ParameterizedTest
    @CsvSource({
            "20.00, 5, 100.00",
            "10.00, 3, 30.00",
            "15.50, 4, 62.00",
            "99.99, 1, 99.99"
    })
    void shouldPlaceOrderWithCorrectTotalPriceParameterized(String unitPrice, String quantity, String expectedTotalPrice) throws Exception {
        var sku = createUniqueSku(SKU);
        var createProductJson = """
                {
                    "id": "%s",
                    "title": "Test Product",
                    "description": "Test Description",
                    "category": "Test Category",
                    "brand": "Test Brand",
                    "price": "%s"
                }
                """.formatted(sku, unitPrice);

        var createProductUri = URI.create(getErpBaseUrl() + "/api/products");
        var createProductRequest = HttpRequest.newBuilder()
                .uri(createProductUri)
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(createProductJson))
                .build();

        var createProductResponse = erpHttpClient.send(createProductRequest, HttpResponse.BodyHandlers.ofString());
        assertThat(createProductResponse.statusCode()).isEqualTo(201);

        var placeOrderJson = """
                {
                    "sku": "%s",
                    "quantity": "%s"
                }
                """.formatted(sku, quantity);

        var placeOrderUri = URI.create(getShopApiBaseUrl() + "/api/orders");
        var placeOrderRequest = HttpRequest.newBuilder()
                .uri(placeOrderUri)
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(placeOrderJson))
                .build();

        var placeOrderResponse = shopApiHttpClient.send(placeOrderRequest, HttpResponse.BodyHandlers.ofString());
        assertThat(placeOrderResponse.statusCode()).isEqualTo(201);

        var placeOrderBody = httpObjectMapper.readTree(placeOrderResponse.body());
        var orderNumber = placeOrderBody.get("orderNumber").asText();

        var viewOrderUri = URI.create(getShopApiBaseUrl() + "/api/orders/" + orderNumber);
        var viewOrderRequest = HttpRequest.newBuilder()
                .uri(viewOrderUri)
                .header("Content-Type", "application/json")
                .GET()
                .build();

        var viewOrderResponse = shopApiHttpClient.send(viewOrderRequest, HttpResponse.BodyHandlers.ofString());
        assertThat(viewOrderResponse.statusCode()).isEqualTo(200);

        var viewOrderBody = httpObjectMapper.readTree(viewOrderResponse.body());
        var expectedTotal = Double.parseDouble(expectedTotalPrice);
        assertThat(viewOrderBody.get("totalPrice").asDouble()).isEqualTo(expectedTotal);
    }

    @Test
    void shouldPlaceOrder() throws Exception {
        var sku = createUniqueSku(SKU);
        var createProductJson = """
                {
                    "id": "%s",
                    "title": "Test Product",
                    "description": "Test Description",
                    "category": "Test Category",
                    "brand": "Test Brand",
                    "price": "20.00"
                }
                """.formatted(sku);

        var createProductUri = URI.create(getErpBaseUrl() + "/api/products");
        var createProductRequest = HttpRequest.newBuilder()
                .uri(createProductUri)
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(createProductJson))
                .build();

        var createProductResponse = erpHttpClient.send(createProductRequest, HttpResponse.BodyHandlers.ofString());
        assertThat(createProductResponse.statusCode()).isEqualTo(201);

        var placeOrderJson = """
                {
                    "sku": "%s",
                    "quantity": "5"
                }
                """.formatted(sku);

        var placeOrderUri = URI.create(getShopApiBaseUrl() + "/api/orders");
        var placeOrderRequest = HttpRequest.newBuilder()
                .uri(placeOrderUri)
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(placeOrderJson))
                .build();

        var placeOrderResponse = shopApiHttpClient.send(placeOrderRequest, HttpResponse.BodyHandlers.ofString());
        assertThat(placeOrderResponse.statusCode()).isEqualTo(201);

        var placeOrderBody = httpObjectMapper.readTree(placeOrderResponse.body());
        var orderNumber = placeOrderBody.get("orderNumber").asText();
        assertThat(orderNumber).startsWith("ORD-");

        var viewOrderUri = URI.create(getShopApiBaseUrl() + "/api/orders/" + orderNumber);
        var viewOrderRequest = HttpRequest.newBuilder()
                .uri(viewOrderUri)
                .header("Content-Type", "application/json")
                .GET()
                .build();

        var viewOrderResponse = shopApiHttpClient.send(viewOrderRequest, HttpResponse.BodyHandlers.ofString());
        assertThat(viewOrderResponse.statusCode()).isEqualTo(200);

        var order = httpObjectMapper.readTree(viewOrderResponse.body());
        assertThat(order.get("orderNumber").asText()).isEqualTo(orderNumber);
        assertThat(order.get("sku").asText()).isEqualTo(sku);
        assertThat(order.get("quantity").asInt()).isEqualTo(5);
        assertThat(order.get("unitPrice").asDouble()).isEqualTo(20.00);
        assertThat(order.get("totalPrice").asDouble()).isEqualTo(100.00);
        assertThat(order.get("status").asText()).isEqualTo("PLACED");
    }
}

